extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts

```diff
@@ -0,0 +1,361 @@
+		if (!url || !username) {
+			return undefined;
+		}
+		let nodes = this.children as ControllerNode[];
+		let index = nodes.findIndex(e => e.url === url && e.username === username);
+		let deleted = undefined;
+		if (index >= 0) {
+			deleted = nodes.splice(index, 1);
+		}
+		return deleted;
+	}
+
+	private getExistingControllerNode(url: string, username: string): ControllerNode {
+		if (!url || !username) {
+			return undefined;
+		}
+		let nodes = this.children as ControllerNode[];
+		return nodes.find(e => e.url === url && e.username === username);
+	}
+}
+
+export class ControllerNode extends ControllerTreeNode {
+	private _url: string;
+	private _username: string;
+	private _password: string;
+	private _rememberPassword: boolean;
+
+	constructor(properties?: {
+		url: string,
+		username: string,
+		password: string,
+		parent: ControllerTreeNode,
+		treeChangeHandler: IControllerTreeChangeHandler,
+		label?: string,
+		description?: string,
+		rememberPassword?: boolean
+	}) {
+		super(Object.assign({
+			label: undefined,
+			nodeType: BdcItemType.controller,
+			iconPath: IconPath.controllerNode
+		}, properties));
+
+		let address = ControllerNode.toIpAndPort(properties.url);
+		this.label = properties.label || `controller: ${address} (${properties.username})`;
+		this.description = properties.description || this.label;
+		this._url = properties.url;
+		this._username = properties.username;
+		this._password = properties.password;
+		this._rememberPassword = !!properties.rememberPassword;
+	}
+
+
+	public async getChildren(): Promise<ControllerTreeNode[]> {
+		if (this.children && this.children.length > 0) {
+			this.clearChildren();
+		}
+
+		if (!this._password) {
+			vscode.commands.executeCommand('bigDataClusters.command.addController', this);
+			return this.children as ControllerTreeNode[];
+		}
+
+		try {
+			let response = await BdcController.getEndPoints(this._url, this._username, this._password, true);
+			if (response && response.endPoints) {
+				let master = response.endPoints.find(e => e.name && e.name === 'sql-server-master');
+				this.addSqlMasterNode(master.endpoint, master.description);
+			}
+			return this.children as ControllerTreeNode[];
+		} catch (error) {
+			let e = error as IControllerError;
+			vscode.window.showErrorMessage(e.message);
+			return this.children as ControllerTreeNode[];
+		}
+	}
+
+	private static toIpAndPort(url: string): string {
+		if (!url) {
+			return;
+		}
+		return url.trim().replace(/ /g, '').replace(/^.+\:\/\//, '').replace(/:(\d+)$/, ',$1');
+	}
+
+	public addSqlMasterNode(endPointAddress: string, description: string): void {
+		let epFolder = this.getEndPointFolderNode();
+		epFolder.addChild(new SqlMasterNode({ endPointAddress, parent: epFolder, treeChangeHandler: this.treeChangeHandler, description }));
+	}
+
+	private getEndPointFolderNode(): FolderNode {
+		let label = 'SQL Servers';
+		let epFolderNode = this.children.find(e => e instanceof FolderNode && e.label === label);
```

Comment:
localize

Reply:
SQL Server is product name. I do not think it should be localized. Do you think 'master' or 'controller' label also should be localized? I am double checking with Kevin.

Reply:
fixed.